tests/subsys/net/lib/nrf_cloud/cloud/src/main.c

```diff
@@ -0,0 +1,2680 @@
+
+	int ret = nrf_cloud_init(&init_param_t);
+
+	zassert_ok(ret, "return should be 0 when init success");
+	zassert_equal(STATE_INITIALIZED, nfsm_get_current_state(),
+		"nrf_cloud lib should be in the initialized state after initialization");
+}
+
+/* Verify that nrf_cloud_init function will fail
+ * when the transport routine fails to initialize
+ */
+ZTEST(nrf_cloud_test, test_init_cloud_transport_fail)
+{
+	struct nrf_cloud_init_param init_param_t = {
+		.event_handler = event_handler,
+		.client_id = NULL,
+	};
+
+	/* Expect an error return value for nct_init */
+	nct_init_fake.custom_fake = fake_nct_init__fails;
+
+	/* Expect other fake to run successfully */
+	nfsm_init_fake.custom_fake = fake_nfsm_init__succeeds;
+	nrf_cloud_codec_init_fake.custom_fake = fake_nrf_cloud_codec_init__succeeds;
+
+	zassert_equal(STATE_IDLE, nfsm_get_current_state(),
+		"nrf_cloud lib should be in the idle state (uninitialized) at the start of test");
+
+	int ret = nrf_cloud_init(&init_param_t);
+
+	zassert_not_equal(0, ret, "return should be non-zero when transport routine init failed");
+	zassert_not_equal(STATE_INITIALIZED, nfsm_get_current_state(),
+		"nrf_cloud lib should not be in the initialized state after init failed");
+}
+
+/* Verify that nrf_cloud_init function will fail
+ * when full modem FOTA update fails to initialize.
+ * Ignored when full modem FOTA update is disabled.
+ */
+ZTEST(nrf_cloud_test, test_init_fmfu_init_failed)
+{
+	Z_TEST_SKIP_IFNDEF(CONFIG_NRF_CLOUD_FOTA_FULL_MODEM_UPDATE);
+	struct dfu_target_fmfu_fdev fmfu_dev_info = {
+		.dev = NULL,
+		.offset = 0L,
+		.size = 0U,
+	};
+
+	struct nrf_cloud_init_param init_param_t = {
+		.event_handler = event_handler,
+		.client_id = NULL,
+		.fmfu_dev_inf = &fmfu_dev_info,
+	};
+
+	/* Expect an error return value for nrf_cloud_fota_fmfu_dev_set */
+	nrf_cloud_fota_fmfu_dev_set_fake.custom_fake = fake_nrf_cloud_fota_fmfu_dev_set__fails;
+
+	/* Expect other fakes to run successfully */
+	nct_init_fake.custom_fake = fake_nct_init__succeeds;
+	nfsm_init_fake.custom_fake = fake_nfsm_init__succeeds;
+	nrf_cloud_codec_init_fake.custom_fake = fake_nrf_cloud_codec_init__succeeds;
+
+	zassert_equal(STATE_IDLE, nfsm_get_current_state(),
+		"nrf_cloud lib should be in the idle state (uninitialized) at the start of test");
+
+	int ret = nrf_cloud_init(&init_param_t);
+
+	zassert_not_equal(0, ret, "return should be non-zero when FOTA init failed");
+	zassert_not_equal(STATE_INITIALIZED, nfsm_get_current_state(),
+		"nrf_cloud lib should not be in the initialized state after init failed");
+}
+
+/* TEST NFSM_SET_CURRENT_STATE_AND_NOTIFY */
+
+/* Verify that nfsm_set_current_state_and notify
+ * will change to the right state that is provided.
+ */
+ZTEST(nrf_cloud_test, test_set_current_state)
+{
+	zassert_equal(STATE_IDLE, nfsm_get_current_state(),
+		"nrf_cloud lib should be in the idle state (uninitialized) at the start of test");
+
+	nfsm_set_current_state_and_notify(STATE_TOTAL, NULL);
```

Comment:
I think you should use a different state for this than `STATE_TOTAL`.  The reason is that it is used to indicate the total count of states, but is otherwise a valid state on its own.  For example, in `nfsm_handle_incoming_event()`, if state >= STATE_TOTAL the function returns -EINVAL.